✨ Commit depth feature

latortuga71 opened the pull request:
#### What kind of change does this PR introduce?
Feature to implement issue to add a flag for` --commit-depth` : https://github.com/ossf/scorecard/issues/2224

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

No `--commit-depth` flag exists to modify number of commits to use in checks.

#### What is the new behavior (if this is a feature change)?**

`--commit-depth` flag is added, which allows the number of commits to be used in checks to be increased/decreased.

- [ ] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

https://github.com/ossf/scorecard/issues/2224

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.

Fixes #

or

NONE
-->

#### Special notes for your reviewer

#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release,
include the string "ACTION REQUIRED".

For more information on release notes see: https://git.k8s.io/release/cmd/release-notes/README.md
-->
```release-note
Added --commit-depth optional flag to allow users to configure the amount of commits scorecard analyzes. 
```

laurentsimon commented on clients/githubrepo/graphql.go:
```diff
@@ -205,19 +212,42 @@ func (handler *graphqlHandler) setup() error {
 		}
-		if err := handler.client.Query(handler.ctx, handler.data, vars); err != nil {
-			handler.errSetup = sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("githubv4.Query: %v", err))
+		if CommitsToAnalyze < 99 {
```
maybe document why 99 is a special number?

latortuga71 replied:
Does this work `// if --number-of-commits flag set to greater than 99 we need to page to query the github api.`?

laurentsimon replied:
SGTM. Maybe say that "graphQL API needs paging for results with size greater than 99"? Not clear whether "we need" is us for our implementation, or a requirement from graphQL

laurentsimon commented on clients/githubrepo/graphql.go:
```diff
@@ -205,19 +212,42 @@ func (handler *graphqlHandler) setup() error {
+		var allCommits []clients.Commit
+		commitsLeft := CommitsToAnalyze
+		vars["commitsToAnalyze"] = githubv4.Int(100)
+		for ; commitsLeft > 0; commitsLeft = commitsLeft - 100 {
```
I wonder if this loop can take care of the case where `CommitsToAnalyze`, instead of having a special case for it.
Same questions for the other places in the code where it's used.

laurentsimon commented on clients/githubrepo/graphql.go:
```diff
@@ -205,19 +212,42 @@ func (handler *graphqlHandler) setup() error {
+		var allCommits []clients.Commit
+		commitsLeft := CommitsToAnalyze
+		vars["commitsToAnalyze"] = githubv4.Int(100)
+		for ; commitsLeft > 0; commitsLeft = commitsLeft - 100 {
```
I would put these in a dedicated function to simplify the code here.

latortuga71 replied:
would something like this make sense ? having two sepererate functions one for paging and one not?
```
func (handler *graphqlHandler) setupWithPaging() {
	commitExpression := handler.commitExpression()
	vars := map[string]interface{}{
		"owner":                  githubv4.String(handler.repourl.owner),
		"name":                   githubv4.String(handler.repourl.repo),
		"pullRequestsToAnalyze":  githubv4.Int(pullRequestsToAnalyze),
		"issuesToAnalyze":        githubv4.Int(issuesToAnalyze),
		"issueCommentsToAnalyze": githubv4.Int(issueCommentsToAnalyze),
		"reviewsToAnalyze":       githubv4.Int(reviewsToAnalyze),
		"labelsToAnalyze":        githubv4.Int(labelsToAnalyze),
		"commitsToAnalyze":       githubv4.Int(CommitsToAnalyze),
		"commitExpression":       githubv4.String(commitExpression),
		"historyCursor":          (*githubv4.String)(nil),
	}
	var allCommits []clients.Commit
	commitsLeft := CommitsToAnalyze
	for vars["commitsToAnalyze"] = githubv4.Int(100); commitsLeft > 0; commitsLeft = commitsLeft - 100 {
		if commitsLeft < 100 {
			vars["commitsToAnalyze"] = githubv4.Int(commitsLeft)
		}
		if err := handler.client.Query(handler.ctx, handler.data, vars); err != nil {
			handler.errSetup = sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("githubv4.Query: %v", err))
			break
		}
		vars["historyCursor"] = handler.data.Repository.Object.Commit.History.PageInfo.EndCursor
		tmp, err := commitsFrom(handler.data, handler.repourl.owner, handler.repourl.repo)
		if err != nil {
			handler.errSetup = err
			break
		}
		allCommits = append(allCommits, tmp...)
	}
	handler.commits = append(handler.commits, allCommits...)
	handler.issues = issuesFrom(handler.data)
	handler.archived = bool(handler.data.Repository.IsArchived)
}


func (handler *graphqlHandler) setupDefault() {
	commitExpression := handler.commitExpression()
	vars := map[string]interface{}{
		"owner":                  githubv4.String(handler.repourl.owner),
		"name":                   githubv4.String(handler.repourl.repo),
		"pullRequestsToAnalyze":  githubv4.Int(pullRequestsToAnalyze),
		"issuesToAnalyze":        githubv4.Int(issuesToAnalyze),
		"issueCommentsToAnalyze": githubv4.Int(issueCommentsToAnalyze),
		"reviewsToAnalyze":       githubv4.Int(reviewsToAnalyze),
		"labelsToAnalyze":        githubv4.Int(labelsToAnalyze),
		"commitsToAnalyze":       githubv4.Int(CommitsToAnalyze),
		"commitExpression":       githubv4.String(commitExpression),
		"historyCursor":          (*githubv4.String)(nil),
	}
	if err := handler.client.Query(handler.ctx, handler.data, vars); err != nil {
		handler.errSetup = sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("githubv4.Query: %v", err))
		return
	}
	handler.archived = bool(handler.data.Repository.IsArchived)
	handler.issues = issuesFrom(handler.data)
	handler.commits, handler.errSetup = commitsFrom(handler.data, handler.repourl.owner, handler.repourl.repo)
	return
}


func (handler *graphqlHandler) setup() error {
	// if --number-of-commits flag set to greater than 99 we need to page to query the github api.
	// else run default setup no paging required.
	if CommitsToAnalyze > 99 {
		handler.setupOnce.Do(func() {
			handler.setupWithPaging()
		})
	} else {
		handler.setupOnce.Do(func() {
			handler.setupDefault()
		})
	}
	return handler.errSetup
}
```

laurentsimon replied:
I'd personally share the code that is shared, and only create a function `populateCommits()` for:
```
for vars["commitsToAnalyze"] = githubv4.Int(100); commitsLeft > 0; commitsLeft = commitsLeft - 100 {
		if commitsLeft < 100 {
			vars["commitsToAnalyze"] = githubv4.Int(commitsLeft)
		}
		if err := handler.client.Query(handler.ctx, handler.data, vars); err != nil {
			handler.errSetup = sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("githubv4.Query: %v", err))
			break
		}
		vars["historyCursor"] = handler.data.Repository.Object.Commit.History.PageInfo.EndCursor
		tmp, err := commitsFrom(handler.data, handler.repourl.owner, handler.repourl.repo)
		if err != nil {
			handler.errSetup = err
			break
		}
		allCommits = append(allCommits, tmp...)
	}
```

Wdut?

latortuga71 replied:
yes that makes sense, something like this ?


```
func populateCommits(handler *graphqlHandler, vars map[string]interface{}) ([]clients.Commit, error) {
	var allCommits []clients.Commit
	commitsLeft := CommitsToAnalyze
	for vars["commitsToAnalyze"] = githubv4.Int(100); commitsLeft > 0; commitsLeft = commitsLeft - 100 {
		if commitsLeft < 100 {
			vars["commitsToAnalyze"] = githubv4.Int(commitsLeft)
		}
		err := handler.client.Query(handler.ctx, handler.data, vars)
		if err != nil {
			return nil, fmt.Errorf("failed to populate commits: %w", err)
		}
		vars["historyCursor"] = handler.data.Repository.Object.Commit.History.PageInfo.EndCursor
		tmp, err := commitsFrom(handler.data, handler.repourl.owner, handler.repourl.repo)
		if err != nil {
			return nil, fmt.Errorf("failed to populate commits: %w", err)
		}
		allCommits = append(allCommits, tmp...)
	}
	return allCommits, nil
}```

laurentsimon replied:
yes, thanks

laurentsimon commented on options/flags.go:
```diff
@@ -132,6 +134,13 @@ func (o *Options) AddFlags(cmd *cobra.Command) {
 		"show extra details about each check",
 	)
 
+	cmd.Flags().IntVar(
```
maybe clarify that it is backward from the HEAD

laurentsimon commented on options/options.go:
```diff
@@ -41,8 +42,8 @@ type Options struct {
 	ResultsFile string
 	ChecksToRun []string
 	Metadata    []string
+	Depth       int
```
just being pedantic: why not call this `number-of-commits`. depth probably work, idk. @olivekl could you advise?

olivekl replied:
+1, `depth` was not self-explanatory to me, `NumberOfCommits` is.

latortuga71 replied:
sure thing, so having it like this `--number-of-commits int   number of commits to check, commits begin backwards from the HEAD` on the command line? and naming the variable in the code NumerOfCommits

laurentsimon replied:
LGTM.
/cc @azeemsgoogle @naveensrinivasan @spencerschrock @raghavkaul no objections?

laurentsimon reviewed (commented):
Maybe add some unit tests.
Also think of adding the results in the final results (pkg/json_results.go) and `pkg/raw_results...`

latortuga71 commented:
> Maybe add some unit tests. Also think of adding the results in the final results (pkg/json_results.go) and `pkg/raw_results...`
@laurentsimon 
for unit tests, where should they be placed? in graphql_e2e_test.go?
or a new file like graphql_test.go i dont see a file that contains tests for other functions like commitsFrom for example.

laurentsimon commented:
> > Maybe add some unit tests. Also think of adding the results in the final results (pkg/json_results.go) and `pkg/raw_results...`
> 
> @laurentsimon for unit tests, where should they be placed? in graphql_e2e_test.go? 

I think this file should be OK. You may need a demo repo that *never* changes so we can keep testing the same thing over time. Shall I create a repo under https://github.com/ossf-tests/? Maybe scorecard-check-graphql-e2e?

latortuga71 commented:
> > > Maybe add some unit tests. Also think of adding the results in the final results (pkg/json_results.go) and `pkg/raw_results...`
> > 
> > 
> > @laurentsimon for unit tests, where should they be placed? in graphql_e2e_test.go?
> 
> I think this file should be OK. You may need a demo repo that _never_ changes so we can keep testing the same thing over time. Shall I create a repo under https://github.com/ossf-tests/? Maybe scorecard-check-ghclient-graphql-e2e?



Sure but before you do that, can you check the following changes i added to the e2e_test.go file. I added some tests for the populateCommits function but ive never used this testing framework so i could have made mistakes. I just added a new context and have my tests in there. Does that work?

```
var _ = Describe("E2E TEST: githubrepo.graphqlHandler", func() {
	var graphqlhandler *graphqlHandler

	BeforeEach(func() {
		graphqlhandler = &graphqlHandler{
			client: graphClient,
		}
	})
        // -> my test below
	Context("E2E TEST: Confirm Paging Commits Works", func() {
		It("Should only have 1 commit", func() {
			_repourl := &repoURL{
				owner:     "ossf",
				repo:      "scorecard",
				commitSHA: clients.HeadSHA,
			}
			_vars := map[string]interface{}{
				"owner":                  githubv4.String("ossf"),
				"name":                   githubv4.String("scorecard"),
				"pullRequestsToAnalyze":  githubv4.Int(1),
				"issuesToAnalyze":        githubv4.Int(30),
				"issueCommentsToAnalyze": githubv4.Int(30),
				"reviewsToAnalyze":       githubv4.Int(30),
				"labelsToAnalyze":        githubv4.Int(30),
				"commitsToAnalyze":       githubv4.Int(1),
				"commitExpression":       githubv4.String("heads/main"),
				"historyCursor":          (*githubv4.String)(nil),
			}
			_ctx := context.Background()
			_logger := log.NewLogger(log.DebugLevel)
			_rt := roundtripper.NewTransport(_ctx, _logger)
			_httpClient := &http.Client{
				Transport: _rt,
			}
			_graphClient := githubv4.NewClient(_httpClient)
			_handler := &graphqlHandler{
				client: _graphClient,
			}
			_handler.init(context.Background(), _repourl)
			commits, err := populateCommits(_handler, _vars)
			Expect(err).To(BeNil())
			Expect(len(commits)).Should(BeEquivalentTo(1))
		})
	})

	Context("E2E TEST: Validate query cost", func() {
		It("Should not have increased for HEAD query", func() {
			repourl := &repoURL{
				owner:     "ossf",
				repo:      "scorecard",
				commitSHA: clients.HeadSHA,
			}
			graphqlhandler.init(context.Background(), repourl)
			Expect(graphqlhandler.setup()).Should(BeNil())
			Expect(graphqlhandler.data).ShouldNot(BeNil())
			Expect(graphqlhandler.data.RateLimit.Cost).ShouldNot(BeNil())
			Expect(*graphqlhandler.data.RateLimit.Cost).Should(BeNumerically("<=", 1))
		})
		It("Should not have increased for commit query", func() {
			repourl := &repoURL{
				owner:     "ossf",
				repo:      "scorecard",
				commitSHA: "de5224bbc56eceb7a25aece55d2d53bbc561ed2d",
			}
			graphqlhandler.init(context.Background(), repourl)
			Expect(graphqlhandler.setup()).Should(BeNil())
			Expect(graphqlhandler.data).ShouldNot(BeNil())
			Expect(graphqlhandler.data.RateLimit.Cost).ShouldNot(BeNil())
			Expect(*graphqlhandler.data.RateLimit.Cost).Should(BeNumerically("<=", 1))
		})
		It("Should not have increased for check run query", func() {
			repourl := &repoURL{
				owner:     "ossf",
				repo:      "scorecard",
				commitSHA: clients.HeadSHA,
			}
			graphqlhandler.init(context.Background(), repourl)
			Expect(graphqlhandler.setupCheckRuns()).Should(BeNil())
			Expect(graphqlhandler.checkData).ShouldNot(BeNil())
			Expect(graphqlhandler.checkData.RateLimit.Cost).ShouldNot(BeNil())
			Expect(*graphqlhandler.checkData.RateLimit.Cost).Should(BeNumerically("<=", 1))
		})
	})
})

```

laurentsimon commented:
LGTM at a high-level. Let's see if the tests pass :)
Fyi, not an expert either on the testing framework :)

latortuga71 commented:
> LGTM at a high-level. Let's see if the tests pass :) Fyi, not an expert either on the testing framework :)

Are there next steps ? 
Also there is an issue with 3 checks specifically the SAST, CI-BestPracticesCheck and CI-Tests Checks. When the number of commits is set to greater than 100, those checks will print a cache miss informational error. due to the setupCheckRuns function being called. And that number not matching the number of commits saved by the client in the setup command.

```
func (handler *graphqlHandler) setupCheckRuns() error {
	handler.setupCheckRunsOnce.Do(func() {
		commitExpression := handler.commitExpression()
		vars := map[string]interface{}{
			"owner":                 githubv4.String(handler.repourl.owner),
			"name":                  githubv4.String(handler.repourl.repo),
			"pullRequestsToAnalyze": githubv4.Int(pullRequestsToAnalyze),
			"commitsToAnalyze":      githubv4.Int(CommitsToAnalyze),
			"commitExpression":      githubv4.String(commitExpression),
			"checksToAnalyze":       githubv4.Int(checksToAnalyze),
		}
		// TODO:
		// sast and ci checks causes cache miss if commits dont match number of check runs.
		// paging for this needs to be implemented if using higher than 100 --number-of-commits
		if CommitsToAnalyze > 99 { // <- paging not supported to maxed out to 99
			vars["commitsToAnalyze"] = githubv4.Int(99)
		}
		if err := handler.client.Query(handler.ctx, handler.checkData, vars); err != nil {
			// quit early without setting crsErrSetup for "Resource not accessible by integration" error
			// for whatever reason, this check doesn't work with a GITHUB_TOKEN, only a PAT
			if strings.Contains(err.Error(), "Resource not accessible by integration") {
				return
			}
			handler.errSetupCheckRuns = err
			return
		}
		handler.checkRuns = parseCheckRuns(handler.checkData)
	})
	return handler.errSetupCheckRuns
}

```



![image](https://user-images.githubusercontent.com/42878263/201957079-bce11fae-4e0d-4936-b1d2-06bfd2f1d0e9.png)

to fix this i think the graphql query for setupcheckruns would need to be rewritten to have paging. But currently it doesnt seem to be a big issue.

azeemshaikh38 reviewed (changes requested):
Sorry about my late review, I was ooo. PR looks great, I have left some comments to be addressed. Once the comments are resolved will work on getting this merged.

Thanks a lot for working on this, this will be a super helpful feature!

latortuga71 commented on options/options.go:
```diff
@@ -23,6 +23,7 @@ import (
 	"github.com/caarlos0/env/v6"
 
 	"github.com/ossf/scorecard/v4/clients"
+	"github.com/ossf/scorecard/v4/clients/githubrepo"
```
That sounds like a good idea, since this effects alot of code, is it ok to make modifications to existing tests to account for the commit depth being passed to InitRepo?

azeemshaikh38 replied:
Yes, for sure. We could have the behavior such that when `commit-depth`/`number-of-commits` is `0` the RepoClient implementation uses some implementation specific default value (30 in case of `githubrepo`). That way, you'd only have to update the existing tests with `InitRepo(..., /*number-of-commits*/ 0)` and the behavior of the tests should remain the same.

latortuga71 replied:
@azeemshaikh38 i just made changes to the InitRepo Function to accept a number of commits, i also added a getcommitdepth func to the interface as i was did not want to pass the arg to pkg.Scorecards func that runs the checks. let me know if this makes sense. some functions in the github client were modified to accept a commit depth var as well.

azeemshaikh38 replied:
Thanks! I would prefer not having getcommitdepth func in the interface, but not a blocker to merge this PR. Would appreciate if you'd remove the func, but feel free to leave it as is and we can clean up later.

latortuga71 replied:
@azeemshaikh38 no problem i will remove it and all instances of 30 will be 0 and handled in InitRepo. will fix tests to reflect that as well.

laurentsimon commented on options/flags.go:
```diff
@@ -59,6 +59,8 @@ const (
 	// FlagFormat is the flag name for specifying output format.
 	FlagFormat = "format"
+
+	FlagNumberOfCommits = "number-of-commits"
```
@olivekl suggested a different name to be more explicit.

azeemshaikh38 replied:
SG

azeemshaikh38 commented:
> Are there next steps ?
> Also there is an issue with 3 checks specifically the SAST, CI-BestPracticesCheck and CI-Tests Checks. When the number of commits is set to greater than 100, those checks will print a cache miss informational error. due to the setupCheckRuns function being called. And that number not matching the number of commits saved by the client in the setup command.

Thanks for pointing this out. Ok to fix this cache miss in a follow-up PR and submit this PR with its current functionality. My only ask would be to leave #2224 open so that we can track this work. Also, if you would like to work on this follow-up too let us know by commenting on #2224, else one of the maintainers or someone else can take this up.

azeemshaikh38 commented on options/options.go:
```diff
@@ -66,6 +66,10 @@ func New() *Options {
 	if opts.LogLevel == "" {
 		opts.LogLevel = DefaultLogLevel
 	}
+	// settings default commit depth of 30 if not passed.
+	if opts.CommitDepth < 0 {
+		opts.CommitDepth = DefaultCommitDepth
+	}
```
Can remove this? This will be handled within RepoClient implementation right?

azeemshaikh38 commented on options/options.go:
```diff
@@ -74,6 +78,8 @@ const (
 	// DefaultCommit specifies the default commit reference to use.
 	DefaultCommit = clients.HeadSHA
 
+	DefaultCommitDepth = 30 // Default Commit Depth
+
```
Remove this too?

azeemshaikh38 commented on clients/localdir/client.go:
```diff
@@ -39,21 +39,26 @@ var (
 		return fmt.Errorf("%w: %v", errInputRepoType, inputRepo)
 	}
-
+	if commitDepth == 0 {
+		client.commitDepth = 30 // default
```
Use `<=` instead to handle the case where someone passes a negative int.

azeemshaikh38 commented on clients/gitlabrepo/client.go:
```diff
@@ -66,9 +67,12 @@ func (client *Client) InitRepo(inputRepo clients.Repo, commitSHA string) error {
 		return sce.WithMessage(sce.ErrRepoUnreachable, err.Error())
 	}
-
+	if commitDepth == 0 {
```
Use `<=`

azeemshaikh38 commented on clients/githubrepo/client.go:
```diff
@@ -69,7 +70,11 @@ func (client *Client) InitRepo(inputRepo clients.Repo, commitSHA string) error {
 		return sce.WithMessage(sce.ErrRepoUnreachable, err.Error())
 	}
-
+	if commitDepth == 0 {
```
Use `<=`

azeemshaikh38 commented on checks/raw/security_policy.go:
```diff
@@ -63,8 +63,9 @@ func SecurityPolicy(c *checker.CheckRequest) (checker.SecurityPolicyData, error)
 	logger := log.NewLogger(log.InfoLevel)
-	dotGitHubClient := githubrepo.CreateGithubRepoClient(c.Ctx, logger)
-	err = dotGitHubClient.InitRepo(c.Repo.Org(), clients.HeadSHA)
+	// HAD TO HARD CODE TO 30
```
Here and everywhere else in the PR, hardcode to use `0` instead of `30`.

azeemshaikh38 reviewed (commented):
Looks good to merge.

latortuga71 commented on clients/githubrepo/graphql.go:
```diff
@@ -229,10 +266,16 @@ func (handler *graphqlHandler) setupCheckRuns() error {
 			"commitExpression":      githubv4.String(commitExpression),
 			"checksToAnalyze":       githubv4.Int(checksToAnalyze),
 		}
+		// TODO:
```
done

latortuga71 commented:
> Looks good to merge.

@azeemshaikh38 let me know if that is ok, all the tests are passing and seems to work fine, i added 0 in all cases where 30 was and used <= instead of ==

i thought i signed the commit, but i guess it didnt so it told me to rebase. that causes conflicts im guessing lmk what i need to do to fix any problems Thanks!

latortuga71 commented:
> > Are there next steps ?
> > Also there is an issue with 3 checks specifically the SAST, CI-BestPracticesCheck and CI-Tests Checks. When the number of commits is set to greater than 100, those checks will print a cache miss informational error. due to the setupCheckRuns function being called. And that number not matching the number of commits saved by the client in the setup command.
> 
> Thanks for pointing this out. Ok to fix this cache miss in a follow-up PR and submit this PR with its current functionality. My only ask would be to leave #2224 open so that we can track this work. Also, if you would like to work on this follow-up too let us know by commenting on #2224, else one of the maintainers or someone else can take this up.

ill probably take a shot a it. may need help will ask in the issue

azeemshaikh38 commented:
> i thought i signed the commit, but i guess it didnt so it told me to rebase. that causes conflicts im guessing lmk what i need to do to fix any problems Thanks!

Thanks. We need to resolve the conflicts before merging. Could you please do that? I'm also looking into it on my side.

latortuga71 commented:
> > i thought i signed the commit, but i guess it didnt so it told me to rebase. that causes conflicts im guessing lmk what i need to do to fix any problems Thanks!
> 
> Thanks. We need to resolve the conflicts before merging. Could you please do that? I'm also looking into it on my side.

Just completed. lmk if there are any issues!

azeemshaikh38 commented:
> > > i thought i signed the commit, but i guess it didnt so it told me to rebase. that causes conflicts im guessing lmk what i need to do to fix any problems Thanks!
> > 
> > 
> > Thanks. We need to resolve the conflicts before merging. Could you please do that? I'm also looking into it on my side.
> 
> Just completed. lmk if there are any issues!

Thanks a lot! Very much appreciate your effort on this PR :) Will work on getting this merged.

gal-legit commented:
FYI the change in pkg/scorecard.go was a breaking change, should've gone into a major semver
